src/main/software/amazon/event/ruler/NameState.java

```diff
@@ -29,22 +31,102 @@ class NameState {
+     *
+     * @return Set of all terminal patterns.
+     */
+    Set<Patterns> getTerminalPatterns() {
+        return patternToTerminalSubRules.keySet();
+    }
+
+    /**
+     * Get all the terminal sub-rules that used a given pattern to lead to this NameState. "Terminal" means the last
+     * field of the sub-rule led to this NameState, and thus, the sub-rule's matching criteria have been satisfied.
+     *
+     * NOTE: This function returns the raw key set from one of NameState's internal maps. Mutating it will corrupt the
+     * state of NameState. Although standard coding practice would warrant returning a copy of the set, or wrapping it
+     * to be immutable, we instead return the raw key set as a performance optimization since this is called repeatedly
+     * on the critical matching path. This avoids creating new data structures and/or copying elements between them.
+     *
+     * @param pattern The pattern that the rules must use to get to this NameState.
+     * @return The sub-rules.
+     */
+    Set<SubRule> getTerminalSubRulesForPattern(Patterns pattern) {
+        return patternToTerminalSubRules.get(pattern);
     }
 
-    // TODO: unit-test these two methods
-    boolean hasRule(final Object name) {
-        return rules.contains(name);
+    /**
+     * Get all the non-terminal sub-rules that used a given pattern to lead to this NameState.
+     *
+     * NOTE: This function returns the raw key set from one of NameState's internal maps. Mutating it will corrupt the
+     * state of NameState. Although standard coding practice would warrant returning a copy of the set, or wrapping it
+     * to be immutable, we instead return the raw key set as a performance optimization since this is called repeatedly
+     * on the critical matching path. This avoids creating new data structures and/or copying elements between them.
+     *
+     * @param pattern The pattern that the rules must use to get to this NameState.
+     * @return The sub-rules.
+     */
+    Set<SubRule> getNonTerminalSubRulesForPattern(Patterns pattern) {
+        return patternToNonTerminalSubRules.get(pattern);
     }
-    void deleteRule(final Object name) {
-        rules.remove(name);
+
+    /**
+     * Get all the sub-rule IDs that match the provided rule, pattern, and isTerminal.
+     *
+     * @param rule The rule that led to this NameState.
+     * @param pattern The pattern used by the given rule to lead to this NameState.
+     * @param isTerminal Whether or not the last field of the rule was used to lead to this NameState.
+     * @return All matching sub-rule IDs.
+     */
+    Set<Double> getSubRuleIds(final Object rule, final Patterns pattern, final boolean isTerminal) {
+        Set<Double> subRuleIds = new HashSet<>();
+        Set<SubRule> subRules = (isTerminal ? patternToTerminalSubRules : patternToNonTerminalSubRules).get(pattern);
+        if (subRules != null) {
+            for (SubRule subRule : subRules) {
```

Comment:
the matching perf will bound to O(n) n is number of sub rules, we might control the sub rule size ...
if the same rule is added multiple times, it should be treated as 1 sub rule with the same id.
I remember one team have the use cases that they cannot efficiently detect the duplicated rule, so just blindly add the rule into Ruler and ask Ruler to dedupe them.

Reply:
Note that this method is only called when deleting a rule. It's the only one in NameState that is O(n). The other NameState methods, including the ones called on the matching code paths, are all O(1).

You raise a good point regarding duplicated rules. We could only re-use existing sub-rule and ID if the entire rule was identical to a previous one. Just knowing that a rule is the same as a previous one up to a current key index is insufficient. So I'm not sure how to tackle this other than having some kind of rule tracker/comparer that we consult before we begin adding a rule. That would have some overhead. Not sure that it's worth it. Also not sure if I'm overlooking a better approach.

I did make a change though to store only the sub-rule ID instead of the whole sub-rule object for non-terminal cases (which is most). This should help a little, but will still lead to the accumulation of IDs if the same rule keeps getting repeatedly added. Let me think about this a bit more.

Reply:
> I remember one team have the use cases that they cannot efficiently detect the duplicated rule

If you remember which team this was, would you mind emailing me and telling me? Thanks.

Reply:
It was Q/A when that team onboarded Ruler, however I cannot access my previous email any more, I guess even it can happen it should be rare in reality, it's better to have a protection cap in case attacker know that and have a loop to repeat add the same rule to downgrade ruler which may affect other customer.

Reply:
I've uploaded a new revision that solves the duplicate rules problem. Note that it involves iterating through all of a NameState's sub-rules for all patterns of a rule's terminal key when adding the rule. In practice, this will rarely be a performance issue, unless many duplicate or near-duplicate rules are added to a machine under different names.